raptor/pipeline.md

```diff
@@ -0,0 +1,103 @@
+- Docs branch should be an input variable
+  - If the metadata is accompanied with new documentation, it should be checked into a branch of `microsoft-graph-docs` repo and branch name should be provided to the pipeline
+- Snippet Generation should be available as a tool to ApiDoctor
+
+## Current State
+
+The tools that are required to answer the question in the [Objectives](#Objectives) section are spread across the following repos.
+1. [microsoftgraph/MSGraph-SDK-Code-Generator](https://github.com/microsoftgraph/MSGraph-SDK-Code-Generator)
+    - Tool to generate C# SDK (Typewriter)
+    - Has a binary release
+    - Can be compiled as a .NET Framework executable
+2. [microsoftgraph/msgraph-sdk-raptor](https://github.com/microsoftgraph/msgraph-sdk-raptor)
+    - Tool to verify whether generated code snippets compile
+    - Creates compilation test cases for each V1 and Beta snippet live at docs.microsoft.com
+    - Consumes Microsoft.Graph V1 and Beta DLLs as Nuget packages
+3. [microsoftgraph/msgraph-sdk-dotnet](https://github.com/microsoftgraph/msgraph-sdk-dotnet)
+    - SDK repo to compile generated code
+4. [microsoftgraph/microsoft-graph-explorer-api](https://github.com/microsoftgraph/microsoft-graph-explorer-api)
+    - Provides an HTTP endpoint to generate C# snippets from HTTP snippets. The HTTP endpoint is the only current option that is exposed for this functionality.
+5. [microsoftgraph/microsoft-graph-docs](https://github.com/microsoftgraph/microsoft-graph-docs)
+    - Contains HTTP snippets
+    - New documentation can be added by forking a branch
+6. [OneDrive/apidoctor](https://github.com/OneDrive/apidoctor)
+    - Tool to extract HTTP snippets from docs and write generated snippet references
+    - Snippet generation functionality is not available in the main repo.
+      - It is available at [andrueastman/apidoctor](https://github.com/andrueastman/apidoctor) / [generate-snippets](https://github.com/andrueastman/apidoctor/tree/generate-snippets).
+      - The functionality is coupled with a PR generation to docs repo upon completion of snippet generation.
+      - Hits public Graph Explorer API HTTP endpoint to generate the snippets
+
+## Proposed Solution
+### Input variables
+1. `version` : `v1.0` or `beta`
+2. either `metadataURL` or metadata artifact from another pipeline run
+3. `docsBranch` : branch on `microsoft-graph-docs` repo, containing new documentation
+4. `typewriterReleaseVersion` : true if the released version of typewriter to be used from Github releases. Otherwise it is expected to be compiled from source.
+
+### Pipeline
+  1. Prepare metadata input
+     1. If metadata is given as a public URL in `metadataURL`, set `metadataPath` variable to that
+        1. Typewriter can take URL as metadata input
+     1. If metadata is an Azure DevOps artifact
+        1. Download the artifact to a local path
+        2. Save that local path as `metadataPath`
+  2. Build snippet generation tool
```

Comment:
i assume this step is short but if not may be a point of optimization since we don't change this frequently.  This could just be a drop share.

Reply:
The reason I want to have build steps is to make sure that we have the opportunity to test changes in individual development branches for each tool.

As an example:
If the pipeline fails due to a fault in snippet generation logic, it can be quickly addressed in a development branch for the tool, can be tested by running the full pipeline by only changing the source for snippet generation tool.

Maybe we can have something similar to `typewriterReleaseVersion` variable to speed things up while consuming the release version.